upf/upf.c

```diff
@@ -48,8 +49,136 @@
 
 #include <vppinfra/tw_timer_1t_3w_1024sl_ov.h>
 
+#if CLIB_DEBUG > 1
+#define upf_debug clib_warning
+#else
+#define upf_debug(...)                          \
+  do { } while (0)
+#endif
+
 static fib_source_t upf_fib_source;
 
+int
+vnet_upf_ue_ip_pool_add_del (u8 * identity, u8 * nwi_name, int is_add)
+{
+  upf_main_t *gtm = &upf_main;
+  upf_ue_ip_pool_info_t *ueip_pool = NULL;
+  uword *p;
+
+  p = hash_get_mem (gtm->ue_ip_pool_index_by_identity, identity);
+
+  if (is_add)
+    {
+      if (p)
+	return VNET_API_ERROR_VALUE_EXIST;
+
+      pool_get (gtm->ueip_pools, ueip_pool);
+      ueip_pool->identity = vec_dup (identity);
+      ueip_pool->nwi_name = vec_dup (nwi_name);
+
+      hash_set_mem (gtm->ue_ip_pool_index_by_identity, identity,
+		    ueip_pool - gtm->ueip_pools);
+
+    }
+  else
+    {
+      if (!p)
+	return VNET_API_ERROR_NO_SUCH_ENTRY;
+
+      ueip_pool = pool_elt_at_index (gtm->ueip_pools, p[0]);
+      hash_unset_mem (gtm->ue_ip_pool_index_by_identity, identity);
+      vec_free (ueip_pool->identity);
+      vec_free (ueip_pool->nwi_name);
+      pool_put (gtm->ueip_pools, ueip_pool);
+    }
+  return 0;
+}
+
+upf_nat_pool_t *
+get_nat_pool_by_name (u8 * name)
+{
+  upf_main_t *gtm = &upf_main;
+  upf_nat_pool_t *np = NULL;
+  uword *p;
+
+  p = hash_get_mem (gtm->nat_pool_index_by_name, name);
+  if (!p)
+    return NULL;
+
+  np = pool_elt_at_index (gtm->nat_pools, p[0]);
+
+  return np;
+}
+
+int
+upf_init_nat_addresses (upf_nat_pool_t * np, ip4_address_t start_addr,
+			ip4_address_t end_addr)
+{
+  u32 i = 0;
+
+  vec_reset_length (np->addresses);
```

Comment:
to avoid unnecessary fragmentation, you could used vec_validate to make sure the vector is large enough to hold all entries before going into the loop.

Reply:
done